test/lit/isorecursive-output-ordering.wast

```diff
@@ -0,0 +1,97 @@
+;; TODO: Autogenerate these checks! The current script cannot handle `rec`.
+
+;; RUN: foreach %s %t wasm-opt -all --hybrid -S -o - | filecheck %s
+
+(module
+ ;; Test that we order groups by average uses and types within each group by uses.
+
+ ;; CHECK:      (rec
+ ;; CHECK-NEXT:  (type $used-a-bit (struct_subtype data))
+ ;; CHECK-NEXT:  (type $unused-6 (struct_subtype data))
+ ;; CHECK-NEXT: )
+
+ ;; CHECK-NEXT: (rec
+ ;; CHECK-NEXT:  (type $used-a-lot (struct_subtype data))
+ ;; CHECK-NEXT:  (type $unused-1 (struct_subtype data))
+ ;; CHECK-NEXT:  (type $unused-2 (struct_subtype data))
+ ;; CHECK-NEXT:  (type $unused-3 (struct_subtype data))
+ ;; CHECK-NEXT:  (type $unused-4 (struct_subtype data))
+ ;; CHECK-NEXT:  (type $unused-5 (struct_subtype data))
+ ;; CHECK-NEXT: )
+
+ (rec
+  (type $unused-1 (struct_subtype data))
+  (type $unused-2 (struct_subtype data))
+  (type $unused-3 (struct_subtype data))
+  (type $unused-4 (struct_subtype data))
+  (type $used-a-lot (struct_subtype data))
+  (type $unused-5 (struct_subtype data))
+ )
+
+ (rec
+  (type $unused-6 (struct_subtype data))
+  (type $used-a-bit (struct_subtype data))
+ )
+
+ (func $use (param (ref $used-a-lot) (ref $used-a-lot) (ref $used-a-lot) (ref $used-a-lot) (ref $used-a-lot) (ref $used-a-lot)) (result (ref $used-a-bit) (ref $used-a-bit) (ref $used-a-bit) (ref $used-a-bit))
+  (unreachable)
+ )
+)
+
+(module
+ ;; Test that we respect dependencies between groups before considering counts.
+
+ ;; CHECK:      (rec
+ ;; CHECK-NEXT:  (type $leaf (struct_subtype  data))
+ ;; CHECK-NEXT:  (type $unused (struct_subtype  data))
+ ;; CHECK-NEXT: )
+
+ ;; CHECK-NEXT: (rec
+ ;; CHECK-NEXT:  (type $used-a-bit (struct_subtype (field (ref $leaf)) data))
+ ;; CHECK-NEXT:  (type $twig (struct_subtype  data))
+ ;; CHECK-NEXT: )
+
+ ;; CHECK-NEXT: (rec
+ ;; CHECK-NEXT:  (type $used-a-lot (struct_subtype  $twig))
+ ;; CHECK-NEXT:  (type $root (struct_subtype  data))
+ ;; CHECK-NEXT: )
+
+ (rec
+  (type $leaf (struct_subtype data))
+  (type $unused (struct_subtype data))
+ )
+
+ (rec
+  (type $twig (struct_subtype data))
+  (type $used-a-bit (struct_subtype (ref $leaf) data))
+ )
+
+ (rec
+  (type $root (struct_subtype data))
+  (type $used-a-lot (struct_subtype $twig))
+ )
```

Comment:
The groups here remain in the same order after sorting. How about flipping them in the input, just to show the change? (otherwise, a stable sort might arrive at the same result)

Reply:
We can't change their order in the input because then the input wouldn't validate. By using each group with different frequencies, I can at least show that the dependencies prevent reordering based on the counts. I'll add an unrelated group in to show more differences as well.